ambry-rest/src/main/java/com.github.ambry.rest/NettyResponseChannel.java

```diff
@@ -305,9 +321,58 @@ void close(Exception exception) {
   private void closeRequest() {
     if (request != null && request.isOpen()) {
       request.close();
+      evaluatePerformanceAndUpdateMetrics();
     }
   }
 
+  private void evaluatePerformanceAndUpdateMetrics() {
+    RestRequestMetricsTracker restRequestMetricsTracker = request.getMetricsTracker();
+    long roundTripTimeInMs = restRequestMetricsTracker.getRoundTripTimeInMs();
+    long timeToFirstBytesInMs = restRequestMetricsTracker.getTimeToFirstByteInMs();
+    long totalOutboundBytes = totalBytesReceived.get();
+    long totalInboundBytes = request.getBytesReceived();
+    RestMethod method = request.getRestMethod();
+    Map<PerformanceIndex, Long> requestPerfToCheck = new HashMap<>();
+    responseStatus = errorResponseStatus == null ? responseStatus : errorResponseStatus;
+    boolean shouldSkipCheck = false;
+    switch (method) {
+      case GET:
+        if (successResponseSet.contains(responseStatus)) {
+          requestPerfToCheck.put(PerformanceIndex.TimeToFirstByte, timeToFirstBytesInMs);
+          requestPerfToCheck.put(PerformanceIndex.AverageBandwidth, totalOutboundBytes * 1000L / roundTripTimeInMs);
+        } else if (nonSuccessResponseSet.contains(responseStatus)) {
+          requestPerfToCheck.put(PerformanceIndex.RoundTripTime, roundTripTimeInMs);
+        }
+        break;
+      case POST:
+        if (successResponseSet.contains(responseStatus)) {
+          requestPerfToCheck.put(PerformanceIndex.AverageBandwidth, totalInboundBytes * 1000L / roundTripTimeInMs);
+        } else if (nonSuccessResponseSet.contains(responseStatus)) {
+          requestPerfToCheck.put(PerformanceIndex.RoundTripTime, roundTripTimeInMs);
+        }
+        break;
+      case PUT:
+      case HEAD:
+      case DELETE:
+        if (successResponseSet.contains(responseStatus) || nonSuccessResponseSet.contains(responseStatus)) {
+          requestPerfToCheck.put(PerformanceIndex.RoundTripTime, roundTripTimeInMs);
+        }
+        break;
+      default:
+        shouldSkipCheck = true;
+        logger.warn("Temporarily no evaluation criteria for " + method + " request. Mark as satisfied directly");
+    }
+    EnumMap<RestMethod, Threshold> thresholdsByMethod =
+        successResponseSet.contains(responseStatus) ? perfConfig.successRequestThresholds
+            : perfConfig.nonSuccessRequestThresholds;
+    if (!shouldSkipCheck && (requestPerfToCheck.isEmpty() || !thresholdsByMethod.get(method)
```

Comment:
So the request is unsatisfied if `requestPerfToCheck` is empty but shouldSkipCheck is false? Is this meant to account for the 5xx error case?

Reply:
Choose to use `responseStatus.isServerError()` to make it clearer.